Expose `RawCommandBuffer::reset()` in `CommandBuffer`

aleksijuvani opened the pull request:
Fixes #issue
PR checklist:
- [ ] `make` succeeds (on *nix)
- [ ] `make reftests` succeeds
- [x] tested examples with the following backends: dx12 vk
- [x] `rustfmt` run on changed code

kvark commented on src/hal/src/command/mod.rs:
```diff
@@ -155,11 +155,20 @@ impl<B: Backend, C, S: Shot, L: Level> CommandBuffer<B, C, S, L> {
+    ///
+    /// The command pool must have been created with the RESET_INDIVIDUAL flag
+    /// to be able to reset individual buffers.
+    pub unsafe fn reset(&mut self, release_resources: bool) {
```
I don't think `reset` makes much sense/use for single-shot buffers. Let's move it under the impls for `Multishot` only?

aleksijuvani replied:
Sure, that sounds reasonable. I've pushed some changes. Is this better?

omni-viral replied:
Single shot buffers are not "one time submit and delete". Rather they should be reset before used again.

omni-viral commented on src/hal/src/command/mod.rs:
```diff
@@ -144,6 +144,17 @@ impl<B: Backend, C> CommandBuffer<B, C, MultiShot, Secondary> {
+    /// commands that have been submitted.
+    ///
+    /// The command pool must have been created with the RESET_INDIVIDUAL flag
+    /// to be able to reset individual buffers.
```
Add also that it must not be in pending state.

kvark commented:
Ah, true. Thanks for correction, @omni-viral!
Looks like the original variant was good.

> On Feb 18, 2019, at 09:36, Zakarum <notifications@github.com> wrote:
> 
> @omni-viral commented on this pull request.
> 
> In src/hal/src/command/mod.rs:
> 
> > @@ -144,6 +144,17 @@ impl<B: Backend, C> CommandBuffer<B, C, MultiShot, Secondary> {
>      }
>  }
>  
> +impl<B: Backend, C, L: Level> CommandBuffer<B, C, MultiShot, L> {
> +    /// Empties a multi-shot command buffer, optionally releasing resources from
> +    /// commands that have been submitted.
> +    ///
> +    /// The command pool must have been created with the RESET_INDIVIDUAL flag
> +    /// to be able to reset individual buffers.
> Add also that it must not be in pending state.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

aleksijuvani commented:
I've changed this back into its original form. I've also added a comment that the command buffer must not be in the "pending" state when it's reset.

I was also under the impression that `OneShot` buffers can only be submitted once and must then be discarded. We should probably address this in the docs for `OneShot` and `MultiShot` so that this is less likely to be misunderstood.

kvark commented:
Your impression was right: reset() is effectively a discard.

Thanks!
Bors r+

> On Feb 18, 2019, at 13:31, Aleksi Juvani <notifications@github.com> wrote:
> 
> I've changed this back into its original form. I've also added a comment that the command buffer must not be in the "pending" state when it's reset.
> 
> I was also under the impression that OneShot buffers can only be submitted once and must then be discarded. We should probably address this in the docs for OneShot and MultiShot so that this is less likely to be misunderstood.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

kvark commented:
Bors retry

> On Feb 18, 2019, at 19:03, bors[bot] <notifications@github.com> wrote:
> 
> Timed out
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
